compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt

```diff
@@ -216,11 +219,23 @@ class ExpressionCodegen(
         if (expression.attributeOwnerId === context.fakeContinuation) {
             addFakeContinuationMarker(mv)
         } else {
-            expression.accept(this, data).materializeAt(type, irType)
+            val promisedValue = expression.accept(this, data)
+            try {
+                promisedValue.materializeAt(type, irType)
+            } catch (e: IllegalAccessException) {
```

Comment:
When you implement some control flow with an exception, it's better (both for clarity and for performance) to introduce a special exception class like this:
```
class CastToNonAccessibleClassException : RuntimeException() {
    override fun fillInStackTrace(): Throwable = this // don't care about stack trace
}
```

Reply:
Didn't know this, thank you!